### ISSUE TITLE
Error in `mo` API

### ISSUE THREAD

**Issue by ashish-2005** [openvinotoolkit/openvino#16095]
So I am converting TensorFlow model in SavedModel format to openvino-IR 
when I use API method
```python
from openvino.tools import mo
model_ir = mo.convert_model(saved_model_dir=str(MODEL_DIR_PATH),data_type='FP32')
```
I got error 
```
Error: Exception occurred during running replacer "REPLACEMENT_ID" (<class 'openvino.tools.mo.load.tf.loader.TFLoader'>): Unexpected exception happened during extracting attributes for node StatefulPartitionedCall/StatefulPartitionedCall/Postprocessor/BatchMultiClassNonMaxSuppression/Slice_3/begin.
Original exception message: index -1 is out of bounds for axis 0 with size 0
```
but when I use command-line  `!mo --saved_model_dir SavedModel --data_type FP32 --output_dir IR` it successfully generates XML and bin file but I have to use API only
can anyone help me out !!

**Comment by rkazants** [comment]
Hi @ashish-2005,

It looks to me that you have some mix in your environment that means in both cases you use our different TF frontends (converts to our opset). Please check IR generated successfully if it contains `name="TensorFlow_Frontend_IR"` inside.
If yes, it means in MO API you use our old frontend (legacy or Python based) and MO tool uses the new one.

Best regards,
Roman

**Comment by ashish-2005** [comment]
Hi @rkazants 
Thanks for your reply 
No the IR generated (XML file) does not contain `name="TensorFlow_Frontend_IR"`
and the version of openVINO is `2022.3.0`

**Comment by rkazants** [comment]
Hi @ashish-2005,

I would like to ask you to check two points:

1. please check that OpenVINO runtime version and MO version match and are the same in both cases. For this, you need to add option `--silent False` as follows:

```
model_ir = mo.convert_model(saved_model_dir=str(MODEL_DIR_PATH),data_type='FP32', silent=False)
```

and 

```
mo --saved_model_dir yolo_v4/ --silent False
```

After that, you can collect these versions for both cases. They look like as follows:

```
OpenVINO runtime version:       xxxx-xxxx
Model Optimizer version:        xxxx-xxxx
```

Please share these versions.

2. Please check that you use the same environment in both cases. Namely, check version of TensorFlow like:

```
import tensorflow as tf
tf.__version__

Best regards,
Roman

**Comment by ashish-2005** [comment]
Hii @rkazants 

I tried API multiple times with `silent=False` and it didn't tell the openVINO versions

for the `mo` command line, the versions are as follows
```
OpenVINO runtime version: 	2022.3.0-9052-9752fafe8eb-releases/2022/3
Model Optimizer version: 	2022.3.0-9052-9752fafe8eb-releases/2022/3
```
TensorFlow version Is `2.9.3`

*(below image is the output of API with silent=False)*
![download](https://user-images.githubusercontent.com/71766106/223496145-92273fef-3a0c-427a-9caf-c6dbda49a358.jpeg)

**Comment by rkazants** [comment]
Hi @ashish-2005,

Thanks for your reply.

Please run Python in in the regular console (not in notebook) and do the following:

```
from openvino.tools import mo
model_ir = mo.convert_model(saved_model_dir=str(MODEL_DIR_PATH),data_type='FP32', silent=False, log_level="DEBUG")
```

it should print out versions of MO and OpenVINO. Also, please share logs after executing it.

Besides, can you please share your model, if possible. The model with fake weights (or light reproducer) will be fine. I just checked if I can convert my yolov4 model using MO API from 2022.3 release. It works on my side.

Best regards,
Roman

**Comment by Deep-Sek** [comment]
Hi, I seem to have a similar issue.
But in my case, my tensorflow version is 2.11.0 and there is a version mismatch as the requirement is tf <= 2.10.
Can you take a look at the complete log below and let me know if that is the reason for this issue or if it could be something else?

If it is the version mismatch, when will support for tf==2.11.0 be release?

```
>>> mo.convert_model(saved_model_dir='Tensorflow/saved_model/', \
...                  transformations_config='/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/tf/ssd_support_api_v2.4.json', \
...                  tensorflow_object_detection_api_pipeline_config='Tensorflow/pipeline.config', \
...                  reverse_input_channels=True, \
...                  output_dir='OpenVino', \
...                  silent=False)
Model Optimizer arguments:
Common parameters:
	- Path to the Input Model: 	None
	- Path for generated IR: 	OpenVino
	- IR output name: 	saved_model
	- Log level: 	ERROR
	- Batch: 	Not specified, inherited from the model
	- Input layers: 	Not specified, inherited from the model
	- Output layers: 	Not specified, inherited from the model
	- Input shapes: 	Not specified, inherited from the model
	- Source layout: 	Not specified
	- Target layout: 	Not specified
	- Layout: 	Not specified
	- Mean values: 	Not specified
	- Scale values: 	Not specified
	- Scale factor: 	Not specified
	- Precision of IR: 	FP32
	- Enable fusing: 	True
	- User transformations: 	Not specified
	- Reverse input channels: 	True
	- Enable IR generation for fixed input shape: 	False
	- Use the transformations config file: 	/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/tf/ssd_support_api_v2.4.json
Advanced parameters:
	- Force the usage of legacy Frontend of Model Optimizer for model conversion into IR: 	False
	- Force the usage of new Frontend of Model Optimizer for model conversion into IR: 	False
TensorFlow specific parameters:
	- Input model in text protobuf format: 	False
	- Path to model dump for TensorBoard: 	None
	- List of shared libraries with TensorFlow custom layers implementation: 	None
	- Update the configuration file with input/output node names: 	None
	- Use configuration file used to generate the model with Object Detection API: 	Tensorflow/pipeline.config
	- Use the config file: 	None
OpenVINO runtime found in: 	/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino
OpenVINO runtime version: 	2022.3.0-9052-9752fafe8eb-releases/2022/3
Model Optimizer version: 	2022.3.0-9052-9752fafe8eb-releases/2022/3
[ WARNING ]  
Detected not satisfied dependencies:
	tensorflow: installed: 2.11.0, required: <= 2.10.0
	fastjsonschema: installed: 2.16.2, required: ~= 2.15.1

Please install required versions of components or run pip installation
pip install openvino-dev[tensorflow]
Traceback (most recent call last):
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/extractor.py", line 434, in extract_node_attrs
    supported, new_attrs = extractor(Node(graph, node))
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/load/tf/loader.py", line 113, in <lambda>
    extract_node_attrs(graph, lambda node: tf_op_extractor(node, check_for_duplicates(tf_op_extractors)))
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/tf/extractor.py", line 82, in tf_op_extractor
    attrs = tf_op_extractors[op](node)
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/common/register_custom_ops.py", line 82, in <lambda>
    lambda node, cls=ex_cls: extension_extractor(
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/common/register_custom_ops.py", line 16, in extension_extractor
    supported = ex.extract(node)
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/tf/const_ext.py", line 19, in extract
    'value': tf_tensor_content(pb_tensor.dtype, shape, pb_tensor),
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/tf/extractors/utils.py", line 106, in tf_tensor_content
    add_value = value_flatten[-1]
IndexError: index -1 is out of bounds for axis 0 with size 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/utils/class_registration.py", line 284, in apply_transform
    replacer.find_and_replace_pattern(graph)
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/load/loader.py", line 14, in find_and_replace_pattern
    self.load(graph)
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/load/tf/loader.py", line 113, in load
    extract_node_attrs(graph, lambda node: tf_op_extractor(node, check_for_duplicates(tf_op_extractors)))
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/front/extractor.py", line 437, in extract_node_attrs
    raise Error(
openvino.tools.mo.utils.error.Error: Unexpected exception happened during extracting attributes for node StatefulPartitionedCall/Postprocessor/BatchMultiClassNonMaxSuppression/Slice_3/begin.
Original exception message: index -1 is out of bounds for axis 0 with size 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/convert.py", line 47, in convert_model
    return _convert(**args)
  File "/home/conex/anaconda3/envs/cmlpipeline/lib/python3.9/site-packages/openvino/tools/mo/convert_impl.py", line 937, in _convert
    raise e.with_traceback(None)
openvino.tools.mo.utils.error.Error: Exception occurred during running replacer "REPLACEMENT_ID" (<class 'openvino.tools.mo.load.tf.loader.TFLoader'>): Unexpected exception happened during extracting attributes for node StatefulPartitionedCall/Postprocessor/BatchMultiClassNonMaxSuppression/Slice_3/begin.
Original exception message: index -1 is out of bounds for axis 0 with size 0
```

**Comment by rkazants** [comment]
Hi @CrackedDS,

Thanks for your message and useful info. Will it be possible to share a link to the model? We need this for further fixing on our side.

Best regards,
Roman

**Comment by ashish-2005** [comment]
Hi @rkazants 

Mine has started working somehow, the only change I made is that I didn't run the `mo` command-line cell (which I was doing before) and now API is doing fine 

As for the model, I am using a pre-trained model [efficientdet-d0](https://tfhub.dev/tensorflow/efficientdet/d0/1) from TensorFlow-Hub
I first loaded the model with `tensorflow_hub` API, then saved it in SavedModel format with `tf.saved_model.save()`  and then used the MO API to create Intermediate-Representation

Thanks again for your time
Ashish

**Comment by ACupofAir** [comment]
> odel.sa

I have the same problem. And I reinstall tensorflow to 2.10.0, unfortunatly, It didn't work

**Comment by ashish-2005** [comment]
> > odel.sa
> 
> I have the same problem. And I reinstall tensorflow to 2.10.0, unfortunatly, It didn't work


Hi @ACupofAir 
I would like you to try reinstalling openVINO with TensorFlow2 `pip install openvino-dev[tensorflow2]`, which is  recommended in the installation guide in case we encounter any compatibility issue

**Comment by ACupofAir** [comment]
> > > odel.sa
> > 
> > 
> > I have the same problem. And I reinstall tensorflow to 2.10.0, unfortunatly, It didn't work
> 
> Hi @ACupofAir I would like you to try reinstalling openVINO with TensorFlow2 `pip install openvino-dev[tensorflow2]`, which is recommended in the installation guide in case we encounter any compatibility issue

I have tried this method, but it didn't work either.

**Comment by rkazants** [comment]
> ing a pre-train

Hi @ashish-2005,

Thanks for the model. Give some time today and I will be back with the results.

Best regards,
Roman

**Comment by CrackedDS** [comment]
> Hi @CrackedDS,
> 
> Thanks for your message and useful info. Will it be possible to share a link to the model? We need this for further fixing on our side.
> 
> Best regards, Roman

Hi @rkazants,
I'm using a modified version of the mobilenet ssd with a custom feature extractor but even the default model is giving the same error. Here's the link to the default model in the model zoo:
http://download.tensorflow.org/models/object_detection/tf2/20200711/ssd_mobilenet_v2_320x320_coco17_tpu-8.tar.gz

**Comment by CrackedDS** [comment]
Hi @rkazants,
 
Is there an update on this? I have a project deadline approaching. If we don't have anything yet, is there any sort of workaround for the moment? Maybe an older version of openvino which doesn't have this issue and is compatible with 2.11 version of tensorflow?

**Comment by rkazants** [comment]
> Hi @rkazants,
> 
> Is there an update on this? I have a project deadline approaching. If we don't have anything yet, is there any sort of workaround for the moment? Maybe an older version of openvino which doesn't have this issue and is compatible with 2.11 version of tensorflow?

Hi @CrackedDS,

I reproduced this issue using both CLI tool and MO API. I sin to freezing functionality (to get frozen `.pb`) for some concrete versions of TF. Some constant node created by freezing turns to be empty. Let me deep dive into why it happens.

Best regards,
Roman

**Comment by ashish-2005** [comment]
Hi @rkazants 

I am facing this [error](https://github.com/openvinotoolkit/openvino/issues/16095#issue-1609683073) again. 
It was working before and stopped working after the recent updates in openvino.
I tried converting other models and it worked completely fine, it's just with this model (EfficientDet) it throws an error I have also tried different versions of this model ie. D1, D2 and encountered the same error.

While this issue resolves, I would like to know is there any other method to convert TF model to OpenVINO-IR other than using the SavedModel format.

Thanks for your time

**Comment by CrackedDS** [comment]
Hi @rkazants,

I remember previous versions of openvino had a way to convert using the checkpoint files directly instead of limiting to the frozen graph and saved model. Is that still an option?

I ask this because, post TF version 2.4 there was a difference in the way the graph is exported. The input nodes are being handled differently. Please check below:
https://github.com/tensorflow/models/commit/0d6ce6025ffc2bed437160fc8b2e9934b3f82fad#diff-b1be120e3a86ccd8af2165d2b081db4c3ab9e0fed6ac6147253f445a303ed746

Since Openvino mo tool has a new json ssd_support_api_v2.4.json for versions specifically after 2.4. I'm thinking maybe in your source code, it doesn't account for this new exporting from tensorflow side?

Could be easy to verify this if there was a way to convert from TF to OpenVino model from checkpoint files instead of saved_model file.

Let me know what you think.

**Comment by CrackedDS** [comment]
Any updates @rkazants?

Cheers
Deepan

**Comment by rkazants** [comment]
@CrackedDS, @ashish-2005, hi guys!
PR: https://github.com/openvinotoolkit/openvino/pull/17311

**Comment by CrackedDS** [comment]
This is great stuff! I’ll pull the latest commit and verify on my end. Thanks! Any idea when this would be released through PyPi even as a prerelease?

**Comment by rkazants** [comment]
> This is great stuff! I’ll pull the latest commit and verify on my end. Thanks! Any idea when this would be released through PyPi even as a prerelease?

@CrackedDS, now we are approaching to 2023.0 release completeness. This fix will be included in this release.
@moslex, please provide more info about release dates if possible.

Best regards,
Roman

**Comment by akladiev** [comment]
This issue will be closed in 2 weeks in case of no activity.

**Comment by ashish-2005** [comment]
This issue has been addressed in PR #17311, hence closing. 
Thanks to all participants
